### PR TITLE
Refacto py4cast start

### PR DESCRIPTION
**Pull request by clemoule**
**implementation of :**
- cli.py for executing the lightning cli
- config_cli.yaml for parsing cli arguments
- a lightningmodule.py which propose a standard structure for lightningmodule
- a datamodule.py which propose a standard structure for datamodule
- a utils.py in which are created utils class that are specialized in one task each. The idea is to create function and override methods methodically during the execution of lightningmodules. (ex: class dummylightningmodule(utils1, utils2, pl.LightningModule) (--> see method resolution order for more details)

- refacto of the old version of lightning.py for autoregressive lightning module/config/cli
- same for a masked auto encoder version (without autoregressive steps)

**A faire:**

(certain)
- DISCUTER DE L'ARBORESCENCE
- enlever dataset_info
- enlever torchdataloadersettings -> batch_size etc...
- rajouter une fonction de save/load checkpoints/weights
- harmoniser les méthodes (named tensor/torch tensor)
- épurer le code
- commenter le code
- on_save/load_checkpoint comme dans lightning.py ?

(peut etre)
- refaire la meme chose pour les num_inter_step
- centraliser les méthodes si possible dans utils.py

pas encore de autoencoderDataModule donc pour l'instant comportement encoder-decoder (x!=y) et pas autoencoder(x=y).

### CODE REVIEW

File: bin/lightning_cli/cli_mae.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from lightning.pytorch.cli import LightningCLI, ArgsType
+from py4cast.lightning_module.mae import MAELightningModule
+from py4cast.data_module.dummy import DummyDataModule
+
+
+def cli_main(args: ArgsType = None):
+    cli = LightningCLI(
+        MAELightningModule,
+        DummyDataModule,
```
Comment (clemoule) [inline bin/lightning_cli/cli_mae.py:9]: pas encore de autoencoderDataModule donc pour l'instant comportement encoder-decoder (x!=y) et pas autoencoder(x=y).

File: py4cast/data_module/dummy.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
+from typing import Union, Dict
+from pathlib import Path
+import lightning.pytorch as pl
+
+# A GARDER
+from py4cast.datasets import get_datasets
+
+# A SUPPRIMER
+from py4cast.datasets.base import TorchDataloaderSettings
```
Comment (clemoule) [inline py4cast/data_module/dummy.py:9]: A terme, enlever TorchDataLoaderSettings -> modifier base.py

**Comment by cbovalo** [comment]
Is this PR ready for review or is it a draft ?
BTW, after a quick review, I find the code would be more readable.

**Comment by clemoule** [comment]
> @clemoule Is this PR ready for review or is it a draft ? BTW, after a quick review, I find the code would be more readable

The code can be executed but i need to verify the plot/log method. We will discuss of the way to redesign the trainers, parsers, lightningmodule and datamodule this afternoon. So this is more of a draft even though it works. We will be set this afternoon.

**Comment by clemoule** [comment]
storing functions in classes (ex: PlotLightningModule) is a bit of over engineering but it helps to group functions and methods by purpose. It makes the code easier to map and understand => quicker to pick up the code => quicker to implement additions to the code and quicker to tinker

**Comment by cbovalo** [comment]
I think it's also a good thing to decouple computing metrics and making plots.
I'll make some comments on the code if improvement is needed.

File: py4cast/lightning_module/dummy.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        self.test_step_metrics_loss = []
+
+        # For making restoring of optimizer state optional (slight hack)
+        self.opt_state = None
+        self.loss = nn.MSELoss(reduction="none")
+        self.metrics = self.get_metrics()
+
+    ###--------------------- MISCELLANEOUS ---------------------###
+
+    def get_metrics(self):
```
Comment (cbovalo) [inline py4cast/lightning_module/dummy.py:69]: Better using the `MetricCollection`class provided by torchmetrics : [MetricCollection](https://lightning.ai/docs/torchmetrics/stable/pages/overview.html#metriccollection)

File: py4cast/settings.py
Comment (cbovalo) [inline py4cast/settings.py:1]: Maybe all these paths could be put in the config file ?

File: py4cast/lightning_module/mae.py
Comment (clemoule) [inline py4cast/lightning_module/mae.py:1]: metrics/loss/step adapted to torch.tensors -> with time, modify to process NamedTensor
